File: 华容道/huarongdao.java
package test;

import java.util.*;

/*
 * 华容道广度优先搜索解法
 * 只能找出最佳路径
 * */
class HuaRongDao {
    //方向定义
    private static final int UP = 0;
    private static final int DOWN = 1;
    private static final int LEFT = 2;
    private static final int RIGHT = 3;
    //终点状态定义
    private static final Integer WIN_STATE = 123456780;
    /* 定义辅助数组
    二维数组，相当于dxdy[4][2],四个数对代表四个方向，方向数值与二维数组第一维相对应，
    dxdy[0][0] = -1;dxdy[0][1] = 0;dxdy[1][0] = 1;dxdy[1][1] = 0 ........
    */
    private static final int[][] dxdy = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    // 3x3的九宫格
    private int[][] arr;

    // 记录空格的位置
    private int x;
    private int y;

    // 定义移动的数组
    private List<Integer> moveArr = new LinkedList<>();
    // 保存已经搜索过的状态
    private Set<Integer> statusSet = new HashSet<>();


    // 代表广搜的每一步，通过lastItem链起来,记录每一步的状态和移动方向并保存上一步的同样的信息
    private class SearchItem {
        private Integer status;
        private Integer dir;
        private SearchItem lastItem;

        SearchItem(Integer status, Integer dir, SearchItem lastItem) {
            this.status = status;
            this.dir = dir;
            this.lastItem = lastItem;
        }

        Integer getStatus() {
            return status;
        }

        Integer getDir() {
            return dir;
        }

        SearchItem getLastItem() {
            return lastItem;
        }
    }

    // 广搜的存储队列
    private List<SearchItem> statusToSearch = new LinkedList<SearchItem>();

    //给定华容道初始值，并遍历找出空格位置(构造方法)
    HuaRongDao() {
        arr = new int[3][3];
        System.out.println("请按照顺序依次输入3*3华容道初始值（0到8且数字不可重复）: ");
        Scanner sc = new Scanner(System.in);
        //键盘录入华容道初始值，对重复录入进行判断，并记录空格位置
        for (int i = 0; i < arr.length; i++) {
            for (int j = 0; j < arr.length; j++) {
                arr[i][j] = sc.nextInt();
                //检验是否重复输入
                for (int l = 0; l <= i; l++) {
                    for (int m = 0; m < j; m++) {     //检验到前一个
                        if (arr[l][m] == arr[i][j]) {
                            System.out.println("输入重复,请重新输入\n");
                            arr[i][j] = sc.nextInt();
                        }
                    }
                    //记录空格位置
                    if (arr[i][j] == 0) {
                        x = i;
                        y = j;
                    }
                }

            }
        }
    }

    // 判断是否可以朝某个方向进行移动
    private boolean canMove(int direction) {
        switch (direction) {
            // y > 0才能左移
            case LEFT:
                return y > 0;
            // y < 2才能右移
            case RIGHT:
                return y < 2;
            // x > 0才能上移
            case UP:
                return x > 0;
            // x < 2才能下移
            case DOWN:
                return x < 2;
        }
        return false;
    }

    // 找出该方向的相反方向
    private int getBackDir(int direction) {
        switch (direction) {
            // y > 0才能左移
            case LEFT:
                return RIGHT;
            // y < 2才能右移
            case RIGHT:
                return LEFT;
            // x > 0才能上移
            case UP:
                return DOWN;
            // x < 2才能下移
            case DOWN:
                return UP;
        }
        return 0;
    }

    // 朝某个方向进行移动，该函数不作判断，直接移动
    // 调用前请自行用canMove先行判断
    private void move(int direction) {
        int temp;
        temp = arr[x + dxdy[direction][0]][y + dxdy[direction][1]];   //方向与第一维数组位置相对应，该位置的数对代表x和y值相应变化，通过加法来实现。注意区分数组位置的0,1,2,3和方向数值的0,1，2,3
        arr[x + dxdy[direction][0]][y + dxdy[direction][1]] = 0;      //也可用 arr[x + dxdy[direction][0]][y + dxdy[direction][1]] = arr[x][y]
        arr[x][y] = temp;
        x = x + dxdy[direction][0];
        y = y + dxdy[direction][1];
    }

    // 某个方向的前进，该函数不作判断，直接移动
    private void moveForward(int direction) {
        move(direction);
        // 该方向记录
        //moveArr.add(direction);                 //步骤记录已经由item记录了，这步看不出有什么意义？？？
    }

    // 某个方向的回退，该函数不作判断，直接移动
    // 其操作和moveForward方法正好相反
    private void moveBack(int direction) {
        move(getBackDir(direction));
        // 记录的移动步骤出栈
        //moveArr.remove(moveArr.size() - 1);      //步骤记录已经由item记录了，这步看不出有什么意义？？？
    }

    // 获取状态，这里把9个数字按顺序组成一个整数来代表状态
    // 方法不唯一，只要能区分九宫格状态就行
    private Integer getStatus() {
        int status = 0;
        for (int i = 0; i < arr.length; i++) {
            for (int j = 0; j < arr.length; j++) {
                status = status * 10 + arr[i][j];
            }
        }
        return status;
    }

    // 根据状态还原九宫格数组
    // 该方法是getStatus的逆过程，感觉没必要用这个方法
    private void recoverStatus(Integer status) {
        for (int i = 0; i < arr.length; i++) {
            for (int j = 0; j < arr.length; j++) {
                arr[2 - i][2 - j] = status % 10;   //逆序赋值，除法取余数
                status = status / 10;       // 除法取整
            }
        }
        getXY();
    }

    // 获取空格的x和y坐标
    private void getXY() {
        for (int i = 0; i < arr.length; i++) {
            for (int j = 0; j < arr.length; j++) {
                if (arr[i][j] == 0) {
                    x = i;
                    y = j;
                }
            }
        }
    }

    // 搜索方法
    private void search() {
        // 如果还有要搜索的状态
        while (statusToSearch.size() > 0) {
            // 队首出列
            SearchItem item = statusToSearch.remove(0);
            Integer status = item.getStatus();
            // 搜到了
            if (status.equals(WIN_STATE)) {
                // 找到路径
                recordRoute(item);
                printRoute();
                return;
            }
            // 根据status还原arr和x，y，不过感觉不到这个方法的意义，画蛇添足
            recoverStatus(status);
            // 4个方向进行遍历
            for (int i = 0; i < 4; i++) {
                // 如果能够朝该方向行走
                if (canMove(i)) {
                    // 向前一步
                    moveForward(i);
                    status = getStatus();

                    // 之前搜过的状态
                    if (statusSet.contains(status)) {
                        moveBack(i);
                        // 放弃
                        continue;   //跳出本次循环，并开始执行下次循环
                    }
                    // 新状态加入待搜索
                    statusSet.add(status);
                    statusToSearch.add(new SearchItem(status, i, item));  //每一步存储节点的信息包含本步的状态以及上一步的节点信息
                    moveBack(i);
                }
            }
        }
    }


    // 根据链表顺藤摸瓜，找到路径
    private void recordRoute(SearchItem item) {
        while (null != item.getLastItem()) {
            moveArr.add(0, item.getDir());
            item = item.getLastItem();
        }
    }

    // 打印路径
    private void printRoute() {
        for (int i = 0; i < moveArr.size(); i++) {
            System.out.print(getDirString(moveArr.get(i)));
            System.out.print(" ");
        }
    }

    // 方向与其对应的字符串
    private String getDirString(int dir) {
        switch (dir) {
            case LEFT:
                return "左";
            case RIGHT:
                return "右";
            case UP:
                return "上";
            case DOWN:
                return "下";
        }
        return null;
    }

    // 打印当前华容道的状态
    private void print() {
        for (int i = 0; i < arr.length; i++) {
            for (int j = 0; j < arr.length; j++) {
                System.out.print(arr[i][j]);
                System.out.print(" ");
            }
            System.out.println();
        }
    }

    // 解题入口方法
    void solve() {
        Integer status = getStatus();
        // 如果已经是胜利状态，返回true
        if (WIN_STATE.equals(status)) {
            return;
        }
        // 初始状态先记录
        statusSet.add(status);
        // 初始状态进入搜索队列
        statusToSearch.add(new SearchItem(status, null, null));
        search();
    }
}

class TsetHuaRongDao {
    public static void main(String[] args) {
        long startTime = System.currentTimeMillis(); //获取开始时间
        HuaRongDao hrd = new HuaRongDao();
        hrd.solve();
        long endTime = System.currentTimeMillis(); //获取结束时间
        long spendTime = endTime - startTime;
        System.out.println("程序运行时间：" + spendTime + "ms"); //输出程序运行时间
    }
}

File: 华容道/huarongdao.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Date  : 2019/11/13 20:49
# @Author: PythonVampire
# @email : vampire@ivamp.cn
# @File  : huarongdao.py
import time


class SearchItem:
    """
    搜索节点
    """

    def __init__(self, status=None, dir=None, last_item=None):
        self.status = status
        self.dir = dir
        self.last_item = last_item
        self.next = None

    def get_status(self): return self.status

    def get_dir(self): return self.dir

    def get_last_item(self): return self.last_item

    def get_next(self): return self.next

    def set_next(self, node): self.next = node


class HuaRongDao2:
    def __init__(self):
        self.up = 0  # 向上
        self.down = 1  # 向下
        self.left = 2  # 向左
        self.right = 3  # 向右

        self.win_state = '123456780'  # 胜利的状态

        self.dxdy = [[-1, 0], [1, 0], [0, -1], [0, 1]]  # 辅助方向数组
        self.arr = [[]]

        self.x = 0  # 空格所在 x 坐标
        self.y = 0  # 空格所在 y 坐标

        self.move_arr = []  # 移动方向记录
        self.status_set = set()  # 保存已经搜索过的状态
        self.status = []  # 当前状态
        self.status_to_search = []  # 广搜储存队列

    def huarongdao2(self, data):
        """
        初始状态输入
        :param data:初始状态
        :return:
        """
        # print('请按照顺序依次输入3*3华容道初始值（0到8且数字不可重复）: ')
        # data = input()
        self.show_status(data, 0)
        l = []
        for item in data:
            l.append(int(item))

        arr = []
        tmp1 = []
        tmp2 = []
        tmp3 = []
        for m in range(0, 3):
            tmp1.append(l[m])
        for m in range(3, 6):
            tmp2.append(l[m])
        for m in range(6, 9):
            tmp3.append(l[m])
        arr.append(tmp1)
        arr.append(tmp2)
        arr.append(tmp3)
        self.arr = arr

        for i in range(len(self.arr)):
            for j in range(len(self.arr[0])):
                if self.arr[i][j] == 0:
                    self.x = i
                    self.y = j

    def can_move(self, direction):
        """
        判断 direction 方向能不能前进
        :param direction: 方向
        :return:
        """
        if direction == self.left:
            return self.y > 0
        elif direction == self.right:
            return self.y < 2
        elif direction == self.up:
            return self.x > 0
        elif direction == self.down:
            return self.x < 2
        else:
            return False

    def get_back_dir(self, direction):
        """
        往 direction 方向后退一步
        :param direction: 方向
        :return:
        """
        if direction == self.left:
            return self.right
        elif direction == self.right:
            return self.left
        elif direction == self.up:
            return self.down
        elif direction == self.down:
            return self.up
        else:
            return False

    def move(self, direction):
        """
        往 direction 方向移动一步
        :param direction: 方向
        :return:
        """
        self.arr[self.x + self.dxdy[direction][0]][self.y + self.dxdy[direction][1]], self.arr[self.x][self.y] = \
            self.arr[self.x][self.y], self.arr[self.x + self.dxdy[direction][0]][self.y + self.dxdy[direction][1]]

        self.x = self.x + self.dxdy[direction][0]
        self.y = self.y + self.dxdy[direction][1]

    def move_forward(self, direction):
        """
        往 direction 方向前进一步
        :param direction: 方向
        :return:
        """
        self.move(direction)

    def move_back(self, direction):
        """
        往 direction 方向退一步, 操作对应前进一步
        :param direction: 方向
        :return:
        """
        self.move(self.get_back_dir(direction))

    def get_status(self):
        """
        获取当前状态
        :return:
        """
        status = ''
        for i in range(len(self.arr)):
            for j in range(len(self.arr[0])):
                x = self.arr[i][j]
                status += str(x)
        return status

    def get_x_y(self):
        """
        获取当前状态下，空格所在位置的 x, y 坐标
        :return:
        """
        for i in range(3):
            for j in range(3):
                if self.arr[i][j] == 0:
                    self.x = i
                    self.y = j

    def recover_status(self, status):
        """
        让 self.arr 恢复到 status 表示的状态
        :param status:状态
        :return:
        """
        for i in range(len(self.arr)):
            for j in range(len(self.arr[0])):
                self.arr[i][j] = int(status[i * 3 + j])

        self.get_x_y()

    def record_route(self, item):
        """
        1. 清空 self.move_arr
        2. 从 item 链表内取出路径
        3. 存放到 self.move_arr 里
        :param item:
        :return:
        """
        self.move_arr = []  # 每次调用的时候清空一次路径记录
        while item.get_last_item():  # 从 item 链表里取每一步的方向，因为是倒序取的，所以取出来也是倒序的
            self.move_arr.append(item.dir)
            item = item.get_last_item()
        self.move_arr = self.move_arr[::-1]  # 做一次倒序，将路径调整为顺序

    def print_route(self):
        """
        转换方向表达方式
        将数字表示的方向，以中文方式打印出来
        :return:
        """
        for dir in self.move_arr:
            print(self.get_dir_string(dir), end=' ')

    def get_dir_string(self, dir):
        """
        数字转中文
        :param dir:
        :return:
        """
        if dir == self.left:
            return '左'
        elif dir == self.right:
            return '右'
        elif dir == self.up:
            return '上'
        elif dir == self.down:
            return '下'
        else:
            return False

    def print(self):
        for i in range(3):
            for j in range(3):
                print(self.arr[i][j])

    @staticmethod
    def show_status(status, offset):
        """
        以 3 x 3 表格的方式打印状态
        :param status: 状态
        :param offset: 显示的偏移量
        :return:
        """
        data = status
        print('当前的状态为: ')
        for i in [0, 3, 6]:
            print(' ' * offset + '{}  {}  {}'.format(data[i], data[i + 1], data[i + 2]))

    def show_dir(self, dir):
        """
        打印方向
        :param dir: 方向
        :return:
        """
        if dir == self.up:
            print(f'方向: 上')
        elif dir == self.down:
            print(f'方向: 下')
        elif dir == self.left:
            print(f'方向: 左')
        elif dir == self.right:
            print(f'方向: 右')
        else:
            print('方向不正确')

    def search(self):
        while self.status_to_search:
            item = self.status_to_search.pop(0)  # 取出队列首位
            status = item.status

            # print('-------------------------------------------------------------------------------------')
            # self.record_route(item)  # 记录下到这个状态的路径
            # self.print_route()
            # print('\t')
            if len(self.move_arr) > 100:  # 如果移动步数大于 1000 则舍弃这个状态，换个状态继续搜索
                continue
            # self.show_status(status, 0)   # 3x3 表格形式显示状态

            if status == self.win_state:
                print('可以胜利, 胜利路径如下: ')
                self.record_route(item)
                self.print_route()
                return

            self.recover_status(status)

            # print('四个方向: ')
            for dir in range(4):  # 4 个方向遍历
                if self.can_move(dir):
                    self.move_forward(dir)
                    status = self.get_status()

                    if status in self.status_set:
                        self.move_back(dir)
                        continue

                    self.status_set.add(status)
                    _item = SearchItem(status, dir, item)
                    self.status_to_search.append(_item)

                    # self.show_dir(dir)    # 显示当前方向
                    # self.show_status(status, 9)   # 以 3x3 表格方式显示状态
                    self.move_back(dir)
        print('无法取得胜利')

    def solve(self):
        status = self.get_status()
        if status == self.win_state:
            return
        self.status_set.add(status)
        self.status_to_search.append(SearchItem(status))
        self.search()


if __name__ == '__main__':
    sc = '341560827'  # 左 左 上 右 下 左 下 右 右 上 左 上 右 下 左 左 下 右 上 左 上 右 下 右 下
    # sc = '413705826'    # 下 左 上 上 右 下 右 下
    # sc = '013425786'    # 右 下 右 下

    start = time.perf_counter()
    hrd = HuaRongDao2()
    hrd.huarongdao2(sc)
    hrd.solve()
    print(time.perf_counter() - start)
